libsolidity/AST.cpp

```diff
@@ -544,10 +566,16 @@ void BinaryOperation::checkTypeRequirements()
 
 void FunctionCall::checkTypeRequirements()
 {
-	m_expression->checkTypeRequirements();
+	// we need to check arguments' type first as their info will be used by m_express(Identifier).
 	for (ASTPointer<Expression> const& argument: m_arguments)
 		argument->checkTypeRequirements();
 
+	auto identifier = dynamic_cast<Identifier*>(m_expression.get());
+	if (identifier)
+		identifier->checkTypeRequirementsWithFunctionCall(*this);
```

Comment:
Another way to solve this would be to store from here inside Expression, which of the three cases we currently have, as it is done for lvalues, and then differentiate inside Identifier::checkTypeRequirements on that.
What do you think about that?